keystore/src/transaction.rs

```diff
@@ -80,379 +75,396 @@ impl EntityId {
+            if deleted_list.contains(&EntityId::from_collection_name(E::COLLECTION_NAME, id)?) {
+                Ok(Some(None))
+            } else {
+                Ok(None)
+            }
         }
     }
 
-    async fn add_operation(&self, op: Operation) {
-        self.operations.write().await.push_back(op);
-    }
-
-    pub async fn save<
-        E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection> + crate::entities::EntityMlsExt,
-    >(
+    pub(crate) async fn find_unique<U: UniqueEntity<ConnectionType = KeystoreDatabaseConnection>>(
         &self,
-        entity: E,
-    ) -> CryptoKeystoreResult<()> {
-        self.add_operation(entity.clone().to_transaction_entity().into()).await;
-        Ok(())
+    ) -> CryptoKeystoreResult<Option<U>> {
+        let cache_result = self.cache.find_unique().await;
+        if let Ok(cache_result) = cache_result {
+            Ok(Some(cache_result))
+        } else {
+            // The deleted list doesn't have to be checked because unique entities don't implement
+            // deletion, just replace. So we can directly return None.
+            Ok(None)
+        }
     }
 
-    pub async fn save_mut<
-        E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection> + crate::entities::EntityMlsExt,
-    >(
+    pub(crate) async fn find_all<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>>(
         &self,
-        mut entity: E,
-    ) -> CryptoKeystoreResult<E> {
-        entity.pre_save().await?;
-        self.add_operation(entity.clone().to_transaction_entity().into()).await;
-        Ok(entity)
+        persisted_records: Vec<E>,
+        params: EntityFindParams,
+    ) -> CryptoKeystoreResult<Vec<E>> {
+        let cached_records: Vec<E> = self.cache.find_all(params.clone()).await?;
+        Ok(self.merge_records(cached_records, persisted_records, params).await)
     }
 
-    pub async fn remove<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>, S: AsRef<[u8]>>(
+    pub(crate) async fn find_many<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>>(
         &self,
-        id: S,
-    ) -> CryptoKeystoreResult<()> {
-        self.add_operation(EntityId::from_collection_name(E::COLLECTION_NAME, id.as_ref())?.into())
-            .await;
-        Ok(())
+        persisted_records: Vec<E>,
+        ids: &[Vec<u8>],
+    ) -> CryptoKeystoreResult<Vec<E>> {
+        let cached_records: Vec<E> = self.cache.find_many(ids).await?;
+        Ok(self
+            .merge_records(cached_records, persisted_records, EntityFindParams::default())
+            .await)
     }
 
-    pub async fn child_groups<
-        E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>
-            + crate::entities::PersistedMlsGroupExt
-            + Sync,
-    >(
+    async fn merge_records<E: crate::entities::Entity<ConnectionType = KeystoreDatabaseConnection>>(
```

Comment:
It would be nice to add a comment describing how the merging is actually performed, i.e. how you get from input records to the output.

Reply:
Thanks for pointing this out.
Added this docstring

```rust
    /// Build a single list of unique records from two potentially overlapping lists.
    /// In case of overlap, records in `records_a` are prioritized.
    /// Identity from the perspective of this function is determined by the output of [crate::entities::Entity::merge_key].
    ///
    /// Further, the output list of records is built with respect to the provided [EntityFindParams]
    /// and the deleted records cached in this [Self] instance.     

```